tools/clang/lib/SPIRV/DeclResultIdMapper.cpp

```diff
@@ -1691,6 +1714,90 @@ class LocationSet {
+/// uses of the same location and component.
+class PackedLocationAndComponentSet {
+public:
+  PackedLocationAndComponentSet(SpirvBuilder &spirvBuilder,
+                                llvm::function_ref<uint32_t(uint32_t)> nextLocs)
+      : spvBuilder(spirvBuilder), assignLocs(nextLocs) {}
+
+  bool assignLocAndComponent(const StageVar *var) {
+    if (tryReuseLocations(var)) {
+      return true;
+    }
+    return assignNewLocations(var);
+  }
+
+private:
+  // Try to pack signature of |var| into unused component slots of already
+  // assigned locations. Returns whether it succeeds or not.
+  bool tryReuseLocations(const StageVar *var) {
+    auto requiredLocsAndComponents = var->getLocationAndComponentCount();
+    for (size_t startLoc = 0; startLoc < nextUnusedComponent.size();
+         startLoc++) {
+      bool canAssign = true;
+      for (uint32_t i = 0; i < requiredLocsAndComponents.location; ++i) {
+        if (startLoc + i >= nextUnusedComponent.size() ||
+            nextUnusedComponent[startLoc + i] +
+                    requiredLocsAndComponents.component >
+                4) {
+          canAssign = false;
+          break;
+        }
+      }
+      if (canAssign) {
+        // Based on Vulkan spec "15.1.5. Component Assignment", a scalar or
+        // two-component 64-bit data type must not specify a Component
+        // decoration of 1 or 3.
+        if (requiredLocsAndComponents.componentAlignment) {
+          reuseLocations(var, startLoc, 2);
+        } else {
+          reuseLocations(var, startLoc, nextUnusedComponent[startLoc]);
+        }
+        return true;
+      }
+    }
+    return false;
+  }
+
+  void reuseLocations(const StageVar *var, uint32_t startLoc,
+                      uint32_t componentStart) {
+    auto requiredLocsAndComponents = var->getLocationAndComponentCount();
+    spvBuilder.decorateLocation(var->getSpirvInstr(), assignedLocs[startLoc]);
+    spvBuilder.decorateComponent(var->getSpirvInstr(), componentStart);
+
+    for (uint32_t i = 0; i < requiredLocsAndComponents.location; ++i) {
+      nextUnusedComponent[startLoc + i] =
+          componentStart + requiredLocsAndComponents.component;
+    }
+  }
+
+  // Pack signature of |var| into new unified stage variables.
+  bool assignNewLocations(const StageVar *var) {
+    auto requiredLocsAndComponents = var->getLocationAndComponentCount();
+    uint32_t loc = assignLocs(requiredLocsAndComponents.location);
+    spvBuilder.decorateLocation(var->getSpirvInstr(), loc);
+
+    uint32_t componentCount = requiredLocsAndComponents.component;
+    for (uint32_t i = 0; i < requiredLocsAndComponents.location; ++i) {
+      assignedLocs.push_back(loc + i);
+      nextUnusedComponent.push_back(componentCount);
+    }
+    return true;
+  }
+
+private:
+  SpirvBuilder &spvBuilder;
+  llvm::function_ref<uint32_t(uint32_t)>
+      assignLocs; ///< A function to assign a new location number.
```

Comment:
nit: Move these comments to their own lines above vars since they don't fit as EOL comments

Reply:
done